Charon/filetypes/CuraPackage.py

```diff
@@ -0,0 +1,124 @@
+
+    ##  Gets a list of paths to machine definition files in the package.
+    def getMachines(self) -> List[str]:
+        return self.listPaths("/definitions")
+
+    ##  Add a new machine definition file.
+    #   \param machine_data The data of the machine definition file in bytes.
+    #   \param package_filename The location to write the data to inside the package.
+    def addMachine(self, machine_data: bytes, package_filename: str) -> None:
+        machine_path_alias = "/definitions"
+        self._ensureRelationExists(virtual_path=machine_path_alias, relation_type="definition", origin="/package.json")
+        self._writeToAlias(machine_path_alias, package_filename, machine_data)
+
+    ##  Gets a list of paths to plugins in the package.
+    def getPlugins(self) -> List[str]:
+        return self.listPaths("/plugins")
+
+    ##  Add a new plugin.
+    #   \param plugin_data The ZIP file containing the plugin files converted to bytes.
+    #   \param plugin_id The ID of the plugin within the package.
+    #   \raises FileNotFoundError If a required file is not in the plugin ZIP file, this error will be raised.
+    def addPlugin(self, plugin_data: bytes, plugin_id: str) -> None:
+        plugin_path_alias = "/plugins"
+        self._ensureRelationExists(virtual_path=plugin_path_alias, relation_type="plugin", origin="/package.json")
+        ignore_string = re.compile("|".join(self.PLUGIN_IGNORED_FILES))
+        required_paths = ["{}/plugin.json".format(plugin_id), "{}/__init__.py".format(plugin_id)]
+        paths_to_add = set()  # type: Set[str]
+
+        # Open the bytes as ZipFile and walk through all the files.
+        with ZipFile(io.BytesIO(plugin_data), "r") as zip_file:
+
+            # Find which files to add.
+            for zip_item in zip_file.filelist:
+                if ignore_string.search(zip_item.filename):
+                    continue
+                paths_to_add.add(zip_item.filename)
+
+            # Validate required files.
+            for required_path in required_paths:
+                if required_path not in paths_to_add:
+                    raise FileNotFoundError("Required file {} not found in plugin directory {}"
+                                            .format(required_path, plugin_id))
+
+            # Add all files.
+            for path in paths_to_add:
+                stream = self.getStream("{}/{}".format(plugin_path_alias, path))
+                stream.write(zip_file.read(path))
+
+    ##  Export the package to bytes.
+    def toByteArray(self, offset: int = 0, count: int = -1) -> bytes:
+        self._validateMetadata()
+        return super().toByteArray(offset, count)
+
+    ##  Creates all the required content types for a .curapackage.
+    def _readContentTypes(self) -> None:
+        super()._readContentTypes()
+        if self._mode != OpenMode.ReadOnly:
+            self.addContentType(extension="xml.fdm_material", mime_type="application/x-ultimaker-material-profile")
```

Comment:
Based on my previous comment regarding "addMaterial" etc handling extensions / mimetype themselves; This should in that case probably be moved so it can actually be accessed)

Reply:
Not sure what you mean here?

Reply:
If you add a material and just set "new material" as filename, it should add the ".xml.fdm_material". But right now the notion of that extension existing at all only exists within this context.

Reply:
Do we have this behavior in other places where libCharon is used?